src/qbe_ir_generator.cpp

```diff
@@ -144,65 +176,104 @@ void QbeIrGenerator::Visit(const DeclStmtNode& decl_stmt) {
+                      FuncScopeTemp{id_num});
+        } else {
+          // 2. int* a = c; c itself stores the address of another integer. We
+          // can directly use the address c currently holds.
+          WriteInstr_("storel {}, {}", FuncScopeTemp{init_num},
+                      FuncScopeTemp{id_num});
+        }
       } else {
-        // 2. int* a = c; c itself stores the address of another integer. We can
-        // directly use the address c currently holds.
-        WriteInstr_("storel {}, {}", FuncScopeTemp{init_num},
+        WriteInstr_("storew {}, {}", FuncScopeTemp{init_num},
                     FuncScopeTemp{id_num});
       }
-    } else {
-      WriteInstr_("storew {}, {}", FuncScopeTemp{init_num},
-                  FuncScopeTemp{id_num});
     }
+    // Set up the number of the id so we know were to load it back.
+    id_to_num[decl.id] = id_num;
   }
-  // Set up the number of the id so we know were to load it back.
-  id_to_num[decl.id] = id_num;
 }
 
 void QbeIrGenerator::Visit(const ArrDeclNode& arr_decl) {
-  int base_addr_num = NextLocalNum();
-  assert(arr_decl.type->IsArr());
-  const auto* arr_type = dynamic_cast<ArrType*>((arr_decl.type).get());
-  auto element_size = arr_type->element_type().size();
-  WriteInstr_("{} =l alloc{} {}", FuncScopeTemp{base_addr_num}, element_size,
-              arr_decl.type->size());
-  id_to_num[arr_decl.id] = base_addr_num;
-
-  for (auto i = std::size_t{0}, e = arr_type->len(); i < e; ++i) {
-    if (i < arr_decl.init_list.size()) {
-      auto& arr_init = arr_decl.init_list.at(i);
-      arr_init->Accept(*this);
+  if (arr_decl.is_global) {
+    const auto* arr_type = dynamic_cast<ArrType*>((arr_decl.type).get());
+    assert(arr_type);
+    if (arr_decl.init_list.size() != 0) {
+      global_var_init_vals.clear();
+      Write_("export data {} = align {} {{ ",
+             user_defined::GlobalPointer{arr_decl.id},
+             arr_type->element_type().size());
+      for (auto i = std::size_t{0}, e = arr_type->len(); i < e; ++i) {
+        auto& arr_init = arr_decl.init_list.at(i);
+        arr_init->Accept(*this);
+        Write_("{}", GenerateQBEInit(global_var_init_vals.at(i)));
+        if (i != e - 1) {
+          Write_(", ");
+        }
+      }
+      Write_(" }}\n");
+    } else {
+      Write_("export data {} = align {} {{ z {} }}\n",
+             user_defined::GlobalPointer{arr_decl.id},
+             arr_type->element_type().size(), arr_decl.type->size());
     }
+  } else {
+    int base_addr_num = NextLocalNum();
+    assert(arr_decl.type->IsArr());
+    const auto* arr_type = dynamic_cast<ArrType*>((arr_decl.type).get());
+    auto element_size = arr_type->element_type().size();
+    WriteInstr_("{} =l alloc{} {}", FuncScopeTemp{base_addr_num}, element_size,
+                arr_decl.type->size());
+    id_to_num[arr_decl.id] = base_addr_num;
+
+    for (auto i = std::size_t{0}, e = arr_type->len(); i < e; ++i) {
```

Comment:
I just realized an issue here: the array should not be filled with 0s if the `init_list` is empty. Is that handled here? If not, it's probably a bug from previous PRs. Addressing it in this PR or a future PR would both be appreciated. 😊

Reply:
Since it's not possible to check the value of uninitialized array, we can check the output of both `qbe` and `llvm` for `test/codegen/array.c`.

```
int main() {
  ...
  int f[2];
  return 0;
}
```
QBE IR
```
export
function w $main() {
@.start.1
@.body.1
        %.147 =w call $printf(l $__builtin_print_format, w %.146)
	%.148 =l alloc4 8
        # no default initialization for array
	%.149 =w copy 0
	ret %.149
}
```

LLVM IR
```
define i32 @main() {
body:
  ...
  %72 = call i32 @printf(ptr @__builtin_print_format, i32 %71)
  %73 = alloca [2 x i32], align 4
  // no default initialization for array
  ret i32 0
}
```